File: apps/services/documents/Documents.Domain/Entities/DocumentAudit.cs
namespace Documents.Domain.Entities;

public sealed class DocumentAudit
{
    public Guid     Id            { get; set; }
    public Guid     TenantId      { get; set; }
    public Guid?    DocumentId    { get; set; }
    public string   Event         { get; set; } = string.Empty;
    public Guid?    ActorId       { get; set; }
    public string?  ActorEmail    { get; set; }
    public string   Outcome       { get; set; } = "SUCCESS";
    public string?  IpAddress     { get; set; }
    public string?  UserAgent     { get; set; }
    public string?  CorrelationId { get; set; }

    /// <summary>Arbitrary JSON detail payload stored as a string (maps to JSON column).</summary>
    public string?  Detail        { get; set; }

    public DateTime OccurredAt    { get; set; }

    // Navigation
    public Document? Document     { get; set; }
}

File: _archived/documents-nodejs/src/domain/entities/document-audit.ts
import type { AuditEventValue } from '@/shared/constants';

/**
 * DocumentAudit — immutable audit trail entry.
 * All writes are INSERT-only; updates and deletes are prohibited at DB level.
 * Designed to be compliance-safe (HIPAA, SOC 2 readiness).
 */
export interface DocumentAudit {
  id:             string;
  tenantId:       string;
  documentId:     string;
  documentVersionId: string | null;
  event:          AuditEventValue;
  actorId:        string;         // userId who performed the action
  actorRoles:     string[];
  correlationId:  string;         // Request trace ID
  ipAddress:      string | null;  // Hashed/truncated — never raw IP in prod
  userAgent:      string | null;
  outcome:        'SUCCESS' | 'DENIED' | 'ERROR';
  detail:         Record<string, unknown>;  // Event-specific metadata
  occurredAt:     Date;
}

export interface CreateAuditInput {
  tenantId:      string;
  documentId:    string;
  documentVersionId?: string;
  event:         AuditEventValue;
  actorId:       string;
  actorRoles:    string[];
  correlationId: string;
  ipAddress?:    string;
  userAgent?:    string;
  outcome:       'SUCCESS' | 'DENIED' | 'ERROR';
  detail?:       Record<string, unknown>;
}
